pyprotista/parsers/ident/msgfplus_2021_03_22_parser.py

```diff
@@ -1,84 +1,106 @@
+            # Short before spec_record begins
+
+        elif entry_tag == (f"{element_tag_prefix}AnalysisSoftware"):
+            version = "msgfplus_" + "_".join(
+                re.findall(r"([/d]*\d+)", entry.attrib["version"])
+            )
 
-    """
-    spectrum = etree.parse(BytesIO(spectrum))
-    spec_records = []
-    spec_level_dict = _get_single_spec_df.reference_dict.copy()
-    spec_level_info = {
-        c.attrib["name"]: c.attrib["value"] for c in spectrum.findall(".//{*}cvParam")
-    }
-    spec_level_dict.update(
-        {
-            _get_single_spec_df.mapping_dict[k]: spec_level_info[k]
-            for k in _get_single_spec_df.mapping_dict
-            if k in spec_level_info
-        }
-    )
-
-    # Iterate children
-    for psm in spectrum.findall(".//{*}SpectrumIdentificationItem"):
-        psm_level_dict = spec_level_dict.copy()
-
-        psm_level_dict.update(
-            {
-                _get_single_spec_df.mapping_dict[k]: psm.attrib[k]
-                for k in _get_single_spec_df.mapping_dict
-                if k in psm.attrib
-            }
-        )
-        cv_param_info = {
-            c.attrib["name"]: c.attrib["value"] for c in psm.findall(".//{*}cvParam")
-        }
-        psm_level_dict.update(
-            {
-                _get_single_spec_df.mapping_dict[k]: cv_param_info[k]
-                for k in _get_single_spec_df.mapping_dict
-                if k in cv_param_info
-            }
-        )
-        user_param_info = {
-            c.attrib["name"]: c.attrib["value"] for c in psm.findall(".//{*}userParam")
+        entry.clear()
+        if entry_tag == (f"{element_tag_prefix}SpectrumIdentificationList"):
+            return version, peptide_lookup, spec_records
+
+
+def _peptide_lookup(entry, entry_tag, sequence, cv_param_modifications, peptide_lookup):
+    if entry_tag == (f"{element_tag_prefix}PeptideSequence"):
+        sequence = {"sequence": entry.text}
+    elif entry_tag == (f"{element_tag_prefix}cvParam"):
+        if entry.attrib["name"] == "unknown modification":
+            cv_param_modifications += entry.attrib["value"] + ":"
+        else:
+            cv_param_modifications += entry.attrib["name"] + ":"
+    elif entry_tag == (f"{element_tag_prefix}Modification"):
+        cv_param_modifications += entry.attrib["location"] + ";"
+    elif entry_tag == (f"{element_tag_prefix}Peptide"):
+        peptide_lookup[entry.attrib["id"]] = {
+            "modifications": cv_param_modifications.rstrip(";")
         }
-        psm_level_dict.update(
-            {
-                _get_single_spec_df.mapping_dict[k]: user_param_info[k]
-                for k in _get_single_spec_df.mapping_dict
-                if k in user_param_info
-            }
-        )
-
-        spec_records.append(psm_level_dict)
-    return pd.DataFrame(spec_records)
+        peptide_lookup[entry.attrib["id"]].update(sequence)
+        cv_param_modifications = ""
+        sequence = ""
```

Comment:
no need to reset - I feel this looks partial very similar to comet parser (dooo - both xml, right ;)) but maybe worth refactoring to avoid code duplication. a mantra worth singing every morning - "Never copy code" :).